File: 7_kyu/Fun_with_binary_numbers.go
package kata

import "math/bits"

// Return list of 0..2^n-1 with all 1 bits in b are also in the list item.
func Solution(n, b int) []int {
	if b <= 0 {
		// for b=0 as defined, and for b<0 as error exit
		return []int{}
	}
	limit := 1 << n
  if limit == 0 {
    panic("out of integer range")
  }
	// accelerate the loop
	step := 1
	for b&step > 0 {
		step <<= 1
	}
	// Each 1 bit of b reduced the result len to 1/2
	accu := make([]int, 0, limit/bits.OnesCount(uint(b)))
	for i := b; i < limit; i += step {
		if i&b == b {
			accu = append(accu, i)
		}
	}
	return accu
}

File: 7_kyu/Fun_with_binary_numbers.rs
// Returns all number up to (2^n)-1 with bit b == 1.
// Supports up to n=31.
fn solution(n: u8, b: u32) -> Vec<u32> {
    let mut accu: Vec<u32> = Vec::new();
    if n == 0 || n > 31 || b == 0 {
        return accu;
    }
    let limit: u32 = 2u32.pow(n as u32);
    let mut i = b;
    while i < limit {
        accu.push(i);
        i = (i + 1) | b;
    }
    accu
}
